feat: add privy server wallet provider

azf20 opened the pull request:
### What changed? Why?
- Introduces a Privy Server Wallet Provider (extending `EvmWalletProvider`), extending the viem provider using @privy-io/server-auth (which has a `createViemAccount` method)
- Follows the "configureWithWallet" pattern (similar to the cdp provider)
- Wallet ID is required
- Supports optional [authorization private key](https://docs.privy.io/guide/server-wallets/authorization/signatures#overview)

#### Qualified Impact
- This is an incremental provider, however it does add @privy-io/server-auth as a dependency. An initial implementation was made using the REST API to minimize dependencies, but adding support for authorization keys and full transaction sending / signing support introduced sufficient complexity that the trade off shifted to adding the dependency - particularly to avoid or easily mitigate breaking changes in future (on either side). However - open to discussion!

John-peterson-coinbase commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,59 @@
+
+    const network = {
+      protocolFamily: "evm" as const,
+      networkId: config.networkId || "84532",
```
network ID for base sepolia is `base-sepolia`

azf20 replied:
thanks, fixed!

John-peterson-coinbase commented on tsconfig.base.json:
```diff
@@ -7,7 +7,7 @@
     "resolveJsonModule": true,
     "esModuleInterop": true,
-    "moduleResolution": "node",
+    "moduleResolution": "node16",
```
why is it required to update `moduleResolution`?

azf20 replied:
I got the following on the `createViemAccount` import using `node`. checking back an additional tsconfig was required as a result (see f49f51d4cf3046fab7aca8813a5f493531df9d2c)
```
Cannot find module '@privy-io/server-auth/viem' or its corresponding type declarations.
  There are types at '/Users/adamfuller/Documents/GitHub/agentkit/node_modules/@privy-io/server-auth/dist/viem.d.ts', but this result could not be resolved under your current 'moduleResolution' setting. Consider updating to 'node16', 'nodenext', or 'bundler'.
```

John-peterson-coinbase commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,59 @@
+import { createWalletClient, http, WalletClient } from "viem";
+import { NETWORK_ID_TO_VIEM_CHAIN } from "../network/network";
+
+interface PrivyWalletConfig {
```
Please add doc string

azf20 replied:
done 🙏

azf20 commented:
thanks @John-peterson-coinbase, added fixes here https://github.com/coinbase/agentkit/pull/242/commits/f49f51d4cf3046fab7aca8813a5f493531df9d2c, including package-lock.json (though on reflection wasn't sure about the policy there). Let me know if there is anything else!

John-peterson-coinbase commented:
Thanks @azf20 - please rebase and sign the commits following the commit signing [instructions](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification)

azf20 commented:
thanks so much, I rebased and verified, hope that looks right!

0xRAG commented:
Hey @azf20 I tested this locally and it's working great! As a couple final asks, could you:
* Add an `examples/langchain-privy-chatbot` – It can be a copy/paste of `examples/langchain-cdp-chatbot` swapping out `CdpWalletProvider` for `PrivyWalletProvider`
* Add an `export_wallet` function (see: [here](https://github.com/coinbase/agentkit/blob/master/typescript/examples/langchain-cdp-chatbot/chatbot.ts#L137-L139)) or otherwise document the best way to save and restore a Privy wallet

Thank you!

azf20 commented:
OK great! I rebased and added an example - I added a comment explaining that Privy server wallets are embedded & controllable via the dashboard & API. Let me know if that makes sense or if you wanted that elsewhere

0xRAG commented on typescript/examples/langchain-privy-chatbot/README.md:
```diff
@@ -0,0 +1,48 @@
+## Requirements
+
+- Node.js 18+
+- [CDP API Key](https://portal.cdp.coinbase.com/access/api)
```
Can replace this with Privy

0xRAG commented:
> OK great! I rebased and added an example - I added a comment explaining that Privy server wallets are embedded & controllable via the dashboard & API. Let me know if that makes sense or if you wanted that elsewhere

Nice, thanks! Two things:
1. Let's add the new example to the [workspaces entry in the root package.json](https://github.com/coinbase/agentkit/blob/master/package.json#L4-L10). Then remove the `package-lock.json` from the new example
2. Thanks for the brief comment explaining Privy Server wallets – I think it would be nice to include it in the example README as well

0xRAG commented:
Also, what are your thoughts on this ask:
> Add an export_wallet function (see: [here](https://github.com/coinbase/agentkit/blob/master/typescript/examples/langchain-cdp-chatbot/chatbot.ts#L137-L139)) or otherwise document the best way to save and restore a Privy wallet

Is this simply saving off the wallet ID, then using that to query Privy?

azf20 commented:
- [x] Added to workspaces entry & removed package.json
- [x] Added some more notes on Privy to the README
- [x] Added an "Export" function for the Wallet Provider which returns the input configuration (App ID, Secret, Wallet ID, Auth Key if applicable) - this isn't the private key, but it's the information required to make any changes on the Privy side
- [x] Fixed commit signing

Thanks!

0xRAG commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,120 @@
+   */
+  exportWallet(): PrivyWalletConfig {
+    return {
+      appId: this.#appId,
+      appSecret: this.#appSecret,
+      walletId: this.#walletId,
+      authorizationKey: this.#authorizationKey,
```
Let's leave off the Privy config stuff and only include the wallet info like the walletId which you already have, and the `networkId` the wallet was created on

azf20 replied:
👍 done, included the auth key too as that is wallet specific (if set)

0xRAG commented on typescript/examples/langchain-privy-chatbot/chatbot.ts:
```diff
@@ -0,0 +1,277 @@
+  const missingVars: string[] = [];
+
+  // Check required variables
+  const requiredVars = ["OPENAI_API_KEY", "PRIVY_APP_ID", "PRIVY_APP_SECRET", "PRIVY_WALLET_ID"];
```
Seeing `PRIVY_WALLET_ID` as required is giving me pause – can we actually update PrivyWalletProvider to create a wallet if it is configured without one? Similar to what we do in CdpWalletProvider: https://github.com/coinbase/agentkit/blob/master/typescript/agentkit/src/wallet-providers/cdpWalletProvider.ts#L147

Sorry didn't bring this up earlier, just thought of it

0xRAG replied:
@azf20 Can you please remove the `PRIVY_WALLET_ID` from the requiredVars?

azf20 replied:
Good catch! Done and updated docs

0xRAG replied:
Thank you! We'll have an update for you soon

0xRAG commented:
@azf20 can you pls rebase `main`. hoping that will help with the CI failures

azf20 commented:
Thanks so much! Rebased + updated README and Changelog

John-peterson-coinbase commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,131 @@
+  /** The ID of the wallet to use */
+  walletId?: string;
+  /** Optional network ID to connect to */
+  networkId?: string;
```
@0xRAG @azf20 - This config should take `chainId` instead of `networkId` . Network ID is a CDP concept and will limit the Privy implementation to only support CDP supported networks. Privy does not have the same constraints as `CdpWalletProvider` and therefore should take `chainId` which will unblock usage of all EVM.

azf20 replied:
Good call! This is done now in https://github.com/coinbase/agentkit/pull/242/commits/ba189e3d19a4cf98c31c86e52db2dacd35daae54

cb-heimdall commented:
Review Error for John-peterson-coinbase @ 2025-02-13 23:31:28 UTC
User failed mfa authentication, public email is not set on your github profile. see [go/mfa-help](http://go/mfa-help)

colfax23 commented on typescript/examples/langchain-privy-chatbot/README.md:
```diff
@@ -0,0 +1,49 @@
+
+This example demonstrates an agent setup as a self-aware terminal style chatbot with a [Privy server wallet](https://docs.privy.io/guide/server-wallets/).
+
+Privy wallets are embedded wallets - learn more at https://docs.privy.io/guide/server-wallets/. The Agentkit integration assumes you have a Privy server wallet ID which you want to use for your agent - creation and management of Privy wallets can be done via the Privy dashboard or API.
```
consider replacing "Privy wallets are embedded wallets" with "Privy's server wallets enable you to securely provision and manage cross-chain wallets via a flexible API"

azf20 replied:
+1

colfax23 reviewed (changes requested):
Wow, this is great!  Thanks for doing this.  I'm Colfax from the Privy team, it's nice to meet you, this was enjoyable to review.

The core change that needs to be made is around how the authorization private key + ID is handled during wallet creation.  Feel free to let me know if you have any questions!

cb-heimdall commented:
Review Error for colfax23 @ 2025-02-14 04:56:41 UTC
User must have write permissions to review

azf20 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,131 @@
+      networkId: config.networkId || "base-sepolia",
+    };
+
+    const chain = NETWORK_ID_TO_VIEM_CHAIN[network.networkId!];
```
done, thanks for the hint!

azf20 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,131 @@
+      (
+        await privy.walletApi.create({
+          chaintype: "ethereum",
+          authorizationKeyIds: config.authorizationKey ? [config.authorizationKey] : undefined,
```
Thanks for that context, sorry I had missed that you need to pass IDs for creation.

I tried 1., but hit an issue where the https://api.privy.io/v1/wallets endpoint was returning a 404 - see this branch on my project repository as a test case https://github.com/azf20/hello-world-computer/commit/cd61b5503eb1908740d27ba655c7f9700a20a59b

So I reverted to 2, but also found that an authorization key was required on new wallet creation, not sure if that is expected? Got the following when I didn't pass in an auth key:
```
 ⨯ Error: Missing `privy-authorization-signature` header.
```
So I updated the error handling and docs accordingly - not an ideal experience but I think ok for now

colfax23 replied:
Good catch - I'll look into this on our end and I think option 2 + pointing folks to the Privy dashboard to create their key is a good approach for now.

azf20 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,131 @@
+  /** Optional network ID to connect to */
+  networkId?: string;
+  /** Optional authorization key for the wallet API */
+  authorizationKey?: string;
```
good call, done

azf20 commented:
Thanks so much for taking a look @colfax23! I fixed and tested new wallet creation, see the approach taken here: https://github.com/coinbase/agentkit/pull/242#discussion_r1956041670. I encountered some issues which might be on the Privy side, so I went ahead with an approach that worked to create new server wallets in Privy:
<img width="943" alt="image" src="https://github.com/user-attachments/assets/eb6260e4-78df-48f0-b6d6-e3723e444c8c" />


Also updated to take any chainId (assuming it's supported by viem), as well as some docs / naming clarifications.

Appreciate the help!

azf20 commented:
An update: @colfax23 shared a URL which will work for wallet creation, and I go that working locally, with the slight downside that it creates a new signing key (with the same public key) on every run. Would be good to clarify expected behaviour on ` ⨯ Error: Missing `privy-authorization-signature` header.`, and whether we prefer the "specify the keyId" or "dynamically create the key" approach here. Thanks!

colfax23 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,165 @@
+   *   appId: "your-app-id",
+   *   appSecret: "your-app-secret",
+   *   walletId: "wallet-id",
+   *   networkId: "base-sepolia"
```
nit: update to `chainId: 84532`

azf20 replied:
good catch!

colfax23 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
````diff
@@ -0,0 +1,165 @@
+   * ```
+   */
+  public static async configureWithWallet(config: PrivyWalletConfig): Promise<PrivyWalletProvider> {
+    const privy = new PrivyClient(config.appId, config.appSecret, {
````
There is an edge case here where the `authorizationPrivateKey` is provided but not `authorizationKeyId` which will get things into a funky state.  Can you do a check to enforce that either both or neither are provided before initializing the `PrivyClient`?

azf20 replied:
I think it's ok to initialise with just an `authorizationPrivateKey` if you're using an existing wallet, not creating a key?

colfax23 replied:
Ah, you're right - if the user has an existing wallet that has a key attached to it, you need to specify `authorizationPrivateKey` and technically don't need to specify the ID.  My bad! Fine to leave as is, good catch.

colfax23 commented on typescript/examples/langchain-privy-chatbot/README.md:
```diff
@@ -0,0 +1,50 @@
+  - PRIVY_APP_ID=
+  - PRIVY_APP_SECRET=
+  - PRIVY_WALLET_ID=[optional, otherwise a new wallet will be created]
+  - PRIVY_WALLET_AUTHORIZATION_PRIVATE_KEY=[optional, only if you are using authorization keys for your server wallets]
```
Can you link the key creation [docs](https://docs.privy.io/guide/server-wallets/authorization/signatures#creation) and / or point folks to the Privy [dashboard](https://dashboard.privy.io/) to create their key + register it with Privy to get their ID?

azf20 replied:
updating README

colfax23 reviewed (approved):
Looks great!  Left a few minor comments, but once addressed this looks good to go from my perspective.

cb-heimdall commented:
Review Error for colfax23 @ 2025-02-14 13:26:16 UTC
User must have write permissions to review

azf20 commented:
Thanks, sounds good - updated READMEs accordingly

John-peterson-coinbase reviewed (changes requested):
LGTM pending a few instances of `networkID` -> `chainID` remaining, a few documentation updates, and a small change to the example.

We can land this in today's EOW release if these are resolved.

@azf20

azf20 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,165 @@
+      privy,
+    });
+
+    const network = {
+      protocolFamily: "evm" as const,
+      chainId: config.chainId || 84532,
+    };
```
Hmm actually thinking do we need this network object? we only use it to fetch the Viem chain, to be passed into the Viem WalletClient - could just pass chainId in directly? the underlying viem provider actually provides the network info https://github.com/coinbase/agentkit/blob/main/typescript/agentkit/src/wallet-providers/viemWalletProvider.ts#L170

azf20 commented on typescript/agentkit/README.md:
```diff
@@ -543,6 +544,47 @@ const walletProvider = new ViemWalletProvider(client, {
+const config = {
+    appId: "PRIVY_APP_ID",
+    appSecret: "PRIVY_APP_SECRET",
+    networkId: "base-sepolia",
```
👍

azf20 commented on typescript/agentkit/README.md:
````diff
@@ -543,6 +544,47 @@ const walletProvider = new ViemWalletProvider(client, {
+
+You can manage authorization keys from your [Privy dashboard](https://dashboard.privy.io/account) or programmatically [using the API](https://docs.privy.io/guide/server-wallets/authorization/signatures).
+
+#### Exporting Privy Wallet information
+
+The `PrivyWalletProvider` can export wallet information by calling the `exportWallet` method. 
+
+```typescript
+const walletData = await walletProvider.exportWallet();
+
+// walletData will be in the following format:
+{
+    walletId: string;
+    authorizationKey: string | undefined;
+    networkId: string;
+}
+```
````
👍

azf20 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,165 @@
+  exportWallet(): {
+    walletId: string;
+    authorizationPrivateKey: string | undefined;
+    networkId?: string;
+  } {
+    return {
+      walletId: this.#walletId,
+      authorizationPrivateKey: this.#authorizationPrivateKey,
+      networkId: this.getNetwork().networkId,
```
👍

azf20 commented on typescript/examples/langchain-privy-chatbot/chatbot.ts:
```diff
@@ -0,0 +1,276 @@
+    const config = {
+      appId: process.env.PRIVY_APP_ID as string,
+      appSecret: process.env.PRIVY_APP_SECRET as string,
+      networkId: process.env.NETWORK_ID || "base-sepolia",
```
👍

azf20 commented on typescript/examples/langchain-privy-chatbot/chatbot.ts:
```diff
@@ -0,0 +1,276 @@
+    process.exit(1);
+  }
+
+  // Warn about optional NETWORK_ID
+  if (!process.env.NETWORK_ID) {
+    console.warn("Warning: NETWORK_ID not set, defaulting to base-sepolia testnet");
+  }
```
👍

azf20 commented on typescript/agentkit/README.md:
```diff
@@ -543,6 +544,47 @@ const walletProvider = new ViemWalletProvider(client, {
+
+#### Authorization Keys
+
+Privy offers the option to use authorization keys to secure your server wallets. When using authorization keys, you must provide the `authorizationPrivateKey` and `authorizationKeyId` parameters to the `configureWithWallet` method.
```
Shuffled it around to put the link to the docs before this 👍

azf20 commented:
thanks so much for the thorough feedback, good catches! pushed changes

0xRAG commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,151 @@
+    let address: `0x${string}`;
+
+    if (!config.walletId) {
+      if (!config.authorizationPrivateKey) {
+        throw new Error("authorizationPrivateKey is required when creating a new wallet");
+      }
+      if (!config.authorizationKeyId) {
+        throw new Error(
+          "authorizationKeyId is required when creating a new wallet with an authorization key, this can be found in your Privy Dashboard",
+        );
```
It's my understanding that an authorization key is optional when creating a Privy server wallet, so can we relax the requirement here?

azf20 replied:
That was also my understanding, though when I tested the header was required - @colfax23 can you confirm if that's only in the case where an account has set up Authorization keys (which mine had)? If so we can update to helpfully handle the resulting error, rather than blocking here

colfax23 replied:
Authorization keys are not required for wallet creation _unless_ you've created one in the dashboard and have given in the permissions `Can create and modify wallets`.  If you've done that, we require it for all wallet creation requests.

Two pieces of feedback we're taking from this & working on:
- This UX of requiring it in all cases if a key is created is not intuitive, we're working on some updates here
- The error message when you try to create a wallet without specifying the key, when it is required, is not very clear (it is `Missing privy-authorization-signature header.`)

For the next steps, I think we can relax this constraint, and @azf20 if you can handle the error gracefully if a key is required but not present that'd be great!

Also, @azf20 when you mention in the README about creating authorization keys in the Privy dashboard, can you add a note to uncheck `Can create and modify wallets` to help prevent people from getting into this state by mistake like you did?  That way they can create a key that is intended for wallet transaction signing purposes only.

0xRAG replied:
Got it, thanks for the clarification @colfax23!

azf20 replied:
thanks both! I added a more helpful error message for that specific case + updated the README

0xRAG commented on typescript/agentkit/README.md:
````diff
@@ -543,6 +544,49 @@ const walletProvider = new ViemWalletProvider(client, {
 });
 ```
 
+### PrivyWalletProvider
````
reminder to add the new sections to the table of contents at the top of the file!

0xRAG commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,162 @@
+      try {
+        const wallet = await privy.walletApi.create({
+          chainType: "ethereum",
+          authorizationKeyIds: [config.authorizationKeyId],
```
@azf20 `config.authorizationKeyId` could be `null` here, we need to optionally pass it:
```
authorizationKeyIds: config.authorizationKeyId ? [config.authorizationKeyId] : undefined,
```

azf20 replied:
good catch, done!

azf20 commented on typescript/agentkit/src/wallet-providers/privyWalletProvider.ts:
```diff
@@ -0,0 +1,162 @@
+    let address: `0x${string}`;
+
+    if (!config.walletId) {
+      if (config.authorizationPrivateKey && !config.authorizationKeyId) {
+        throw new Error(
+          "authorizationKeyId is required when creating a new wallet with an authorization key, this can be found in your Privy Dashboard",
+        );
+      }
```
makes sense, done!

0xRAG reviewed (approved):
Amazing work, thanks for doing this @azf20!

@colfax23 thank you as well for your review!

LGTM, @John-peterson-coinbase can you take a final pass?

🚀

John-peterson-coinbase reviewed (approved):
🚀
